modules/thermalmanager.c

```diff
@@ -52,545 +156,995 @@
+/** Broadcast overall device thermal status via D-Bus
+ *
+ * Broadcast the changes as D-Bus signal
+ *   com.nokia.thermalmanager.thermal_state_change_ind(state_name)
+ *
+ * @param status          device overall thermal status
+ */
+static void
+thermal_manager_broadcast_status_dbus(THERMAL_STATUS status)
 {
-    is_in_ta_test = true;
-    dsme_log(LOG_NOTICE, "thermalmanager: set TA test mode");
+    static THERMAL_STATUS prev = THERMAL_STATUS_NORMAL;
+
+    if( prev == status )
+        goto EXIT;
+
+    prev = status;
+
+    const char *arg = thermal_status_name(status);
+
+    dsme_log(LOG_NOTICE, PFIX"send dbus signal %s.%s(%s)", dbus_interface,
+             thermalmanager_state_change_ind, arg);
+
+    DsmeDbusMessage *sig =
+        dsme_dbus_signal_new(dbus_object,
+                             dbus_interface,
+                             thermalmanager_state_change_ind);
+
+    dsme_dbus_message_append_string(sig, arg);
+    dsme_dbus_signal_emit(sig);
+
+EXIT:
+    return;
 }
-#endif
-
-
-// TODO: rename module_fn_info_t to dsme_binding_t
-module_fn_info_t message_handlers[] = {
-  DSME_HANDLER_BINDING(DSM_MSGTYPE_WAKEUP),
-  DSME_HANDLER_BINDING(DSM_MSGTYPE_DBUS_CONNECT),
-  DSME_HANDLER_BINDING(DSM_MSGTYPE_DBUS_DISCONNECT),
-#ifdef DSME_THERMAL_TUNING
-  DSME_HANDLER_BINDING(DSM_MSGTYPE_SET_TA_TEST_MODE),
-#endif
-  { 0 }
-};
 
+/** Request sensor status update
+ *
+ * All thermal objects have name matching the one given are
+ * updated.
+ *
+ * For example, if there are sensors "core0", "core1", "core2" and "core3"
+ *
+ * Requesting "core0" update does "core0" only.
+ *
+ * But requesting "core" update does all four.
+ *
+ * @param sensor_name  sensor name / sensor group name prefix
+ */
+bool
+thermal_manager_request_sensor_update(const char *sensor_name)
+{
+    bool ack = false;
+    for( GSList *item = thermal_objects; item; item = item->next ) {
+        thermal_object_t *object = item->data;
+        if( !thermal_object_has_name_like(object, sensor_name) )
+            continue;
+        thermal_object_request_update(object);
+        ack = true;
+        break;
+    }
+    return ack;
+}
 
-void module_init(module_t* handle)
+/** Check for pending sensor status request
+ *
+ * Check if any thermal objects with name matching the given on
+ * are still waiting for status update.
+ *
+ * For example, if there are sensors "core0", "core1", "core2" and "core3"
+ *
+ * Requesting "core0" checks "core0" only.
+ *
+ * But requesting "core" checks all four.
+ *
+ * @param sensor_name  sensor name / sensor group name prefix
+ */
+static bool
+thermal_manager_have_pending_sensor_update(const char *sensor_name)
 {
-  dsme_log(LOG_DEBUG, "thermalmanager.so loaded");
+    bool pending = false;
+
+    for( GSList *item = thermal_objects; item; item = item->next ) {
+        thermal_object_t *object = item->data;
+
+        /* Must be waiting for status */
+        if( !thermal_object_update_is_pending(object) )
```

Comment:
Maybe do the name checks before the pending checks? Or is the pending check guaranteed to be "cheaper" (`continue` early) than the name comparison?